### PR TITLE
[8.0] Add stock_scanner module

### PR DESCRIPTION
**Pull request by Garamotte**
This module allows to write scenarios, to be used (mainly) from smart barcode scanners.
The scenarios can be created directly from Odoo, or added as any other data in a module.
A specific odoo client (written using python/ncurses) is then used to run the scenarios from the barcode scanner (usually through SSH).

A branch for version 9.0 is also available (only an XML ID in the standard stock module changes between 8.0 an 9.0).

### CODE REVIEW

File: stock_scanner/.description [hunk 1]
```diff
@@ -0,0 +1 @@
+Allows managing barcode readers with simple scenarios
```
Comment (elicoidal) [inline stock_scanner/.description]: remove file?

File: stock_scanner/.oerp_sentinelrc [hunk 2]
```diff
@@ -0,0 +1,5 @@
+[openerp]
```
Comment (elicoidal) [inline stock_scanner/.oerp_sentinelrc]: remove file?

File: stock_scanner/README.rst [hunk 3]
```diff
@@ -0,0 +1,187 @@
+====================================================
```
Comment (elicoidal) [inline stock_scanner/README.rst:5]: Hi @sylvain-garancher Thanks for your contribution
Please follow the OCA README template:
https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst

File: stock_scanner/README.rst [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+Stock Scanner : WorkFlow engine for scanner hardware
+====================================================
+
+Allows managing barcode readers with simple scenarios
+- You can define a workfow for each object (stock picking, inventory, sale, etc)
```
Comment (elicoidal) [inline stock_scanner/README.rst:11]: rst list

File: stock_scanner/README.rst [hunk 5]
```diff
@@ -0,0 +1,187 @@
+====================================================
+Stock Scanner : WorkFlow engine for scanner hardware
+====================================================
+
+Allows managing barcode readers with simple scenarios
```
Comment (elicoidal) [inline stock_scanner/README.rst]: colon

File: stock_scanner/README.rst [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+- You can define a workfow for each object (stock picking, inventory, sale, etc)
+- Works with all scanner hardware model (just SSH client required)
+
+The "sentinel" specific ncurses client, available in the "hardware" directory, requires the "openobject-library" python module, available from pip :
```
Comment (elicoidal) [inline stock_scanner/README.rst]: move to installation section

File: stock_scanner/README.rst [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+    $ sudo pip install openobject-library
+
+Some demo/tutorial scenarios are available in the "demo" directory of the module.
+These scenarios, are automatically imported when installing a new database with demo data.
+
+How does it work ?
```
Comment (elicoidal) [inline stock_scanner/README.rst]: Usage

File: stock_scanner/README.rst [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+==================
+
+On start-up, the client lists available scenarii.
+When the user selects a scenario, the current scenario and step are stored on the hardware configuration's entry in Odoo.
```
Comment (elicoidal) [inline stock_scanner/README.rst]: s/step/steps

File: stock_scanner/README.rst [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+
+When the client sends a message to the server, the next step is selected depending on the current step and the message sent.
+Then, the server returns the result of the step, which contains the step type code and the text to display on the hardware screen.
```
Comment (elicoidal) [inline stock_scanner/README.rst]: The server then returns the result of the step,[..] its type code

File: stock_scanner/README.rst [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+
+You have to declare some hardware scanners in Odoo.
+
+Go to "Warehouse > Configuration > Scanner Hardware" and create a new record.
+
+The "step type code" sent by the "sentinel.py" client at start-up is the IP address of the hardware, if connected through SSH.
+
+If needed enable Login/Logout
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+The module come with 2 predifined scenarii for Login and Logout. The functionality is disabled by default and the user to use in
```
Comment (elicoidal) [inline stock_scanner/README.rst:50]: fix all text following PEP8

File: stock_scanner/README.rst [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+scenario no more appear. 
+
+The Login/logout functionality enable you to specify on the scenario a list of users and/or a list of groups with access to the scenario.
+
+To enable the Login/logout functionality:
+    * Go to "Settings > Warehouse" and check the checkbox Login/logout scenarii enabled.
+    * Create a *Technical User* 'sentinel' **without roles in Human Resources** and with 'Sentinel: technical users' checked.
+    * Use this user to launch your sentinel session.
+
+Be carefull, the role *Sentinel: technical users* is a technical role and should be used only by sentinel.
```
Comment (elicoidal) [inline stock_scanner/README.rst]: careful
should only be used

File: stock_scanner/README.rst [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+
+The timeout of sessions is managed by a dedicated cron that reset the incative sessions. The timeout can be configured on
```
Comment (elicoidal) [inline stock_scanner/README.rst]: indicative

File: stock_scanner/__init__.py [hunk 13]
```diff
@@ -0,0 +1,30 @@
+# -*- coding: utf-8 -*-
```
Comment (elicoidal) [inline stock_scanner/__init__.py:1]: Please use simple headers following OCA templates
https://github.com/OCA/maintainer-tools/blob/master/template/module/__openerp__.py#L3

File: stock_scanner/__openerp__.py [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+{
```
Comment (elicoidal) [inline stock_scanner/__openerp__.py:5]: Please follow the OCA README template:
https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst
Add OCA author
Version 8.0.1.0.0
Remove empty keys

File: stock_scanner/hardware/sentinel.py [hunk 15]
```diff
@@ -0,0 +1,888 @@
...
+
+class SentinelException (Exception):
+    pass
+
+
+class SentinelBackException (SentinelException):
+    pass
+
+if __name__ == '__main__':
+    curses.wrapper(Sentinel)
```
Comment (elicoidal) [inline stock_scanner/hardware/sentinel.py:867]: Update images to v8

File: stock_scanner/models/scanner_scenario_transition.py [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+        string='Sequence',
+        default=0,
+        required=False,
+        help='Sequence order')
+    from_id = fields.Many2one(
+        comodel_name='scanner.scenario.step',
+        string='From',
+        required=True,
+        ondelete='cascade',
+        help='Step which launches this transition')
```
Comment (elicoidal) [inline stock_scanner/models/scanner_scenario_transition.py]: Add periods at the end of all help sentences

File: stock_scanner/scripts/export_scenario.py [hunk 17]
```diff
@@ -0,0 +1,259 @@
...
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU General Public License for more details.
+#
+#    You should have received a copy of the GNU General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+import re
+from oobjlib.connection import Connection
```
Comment (elicoidal) [inline stock_scanner/scripts/export_scenario.py:8]: shouldnot we add exception handling for the import?

**Comment by elicoidal** [comment]
That's an impressive contribution!
I wish all modules were that documented :)

**Comment by Garamotte** [comment]
Thanks for this first review Eric.
I updated `README.rst` and all licences in files to follow OCA guidelines and fix typos.
Periods have been added on all help sentences.
The `.description` file has been removed (included in the `summary` key in `__openerp__.py`).
The `.oerp_sentinelrc` file has been removed (included in the relevant part of `README.rst`.

**Comment by laetitia-gangloff** [comment]
:+1:  (functional)

We use this module since 7.0. And we have also a version 8.0 in production. All works as expected.
Thank you !

**Comment by elicoidal** [comment]
Great! Thanks again for your contribution

**Comment by clonedagain** [comment]
Maybe you could leave the .mo file out. They're complied from .po files I suppose?

File: stock_scanner/hardware/sentinel.py [hunk 18]
```diff
@@ -0,0 +1,866 @@
...
+    Manages scanner terminals
+    """
+
+    def __init__(self, stdscr):
+        """
+        Initialize the sentinel program
+        """
+        # Read user configuration
+        config = ConfigParser.SafeConfigParser(DEFAULT_CONFIG)
+        config.read('.oerp_sentinelrc')
```
Comment (clonedagain) [inline stock_scanner/hardware/sentinel.py]: Maybe it could be named `.openerp_sentinelrc`, given that Odoo still uses `.openerp_serverrc`?

**Comment by lmignon** [comment]
:+1:  We use this module for many of our clients. Through this, they are able to perform all procedures for the management of a stock through scanner hardware (ex Motorala MC3190 winCE, MC3200 android). Moreover the approach of emulating and controlling  the user interface in a terminal eases the deployments because once the scanner hardware are configured for ssh access to a terminal running 'sentinel' there is no more need to deploy something on the scanner hardware. For the record, our client is able to make a full inventory into 2 warehouses with more than 10.000 locations in less than a few hours. The inventory is done by 30 employees each equipped with a scanner and operating in parallel.

File: stock_scanner/hardware/datalogic/PowerScan/PowerscanService.py [hunk 19]
```diff
@@ -0,0 +1,219 @@
...
+## quantity = ['Nombre de carton','','',20]
+## checkquantity(quantity,quantity[3])
+## yesno = ['Confirmez','','','']
+## yesnochoice(yesno)
+##
+## x.cr()
+##closeserial()
+##exit()
+##x.greenledon()
+##x.cursorcr()
```
Comment (clonedagain) [inline stock_scanner/hardware/datalogic/PowerScan/PowerscanService.py]: Please remove this, it's in git history.

File: stock_scanner/hardware/datalogic/PowerScan/PowerscanService.py [hunk 20]
```diff
@@ -0,0 +1,219 @@
...
+print 'Welcome to PowerScan Sentinel\n\n  '
+
+while True:
+    (hwd, msg) = pws.read()
+    print '%s -> %s' % (hwd, msg)
+    pws.cleardisplay(hwd)
+    pws.write("-> %s" % msg)
+    pws.cr()
+    menu = ['Inventaire', 'Reception', 'Preparation', 'Controle', 'Terminer']
+    #displaymenu(PS,menu,0,0)
```
Comment (clonedagain) [inline stock_scanner/hardware/datalogic/PowerScan/PowerscanService.py]: Please remove this

File: stock_scanner/hardware/datalogic/PowerScan/PowerscanService.py [hunk 21]
```diff
@@ -0,0 +1,219 @@
...
+    [(getattr(signal, sign), sign) for sign in LST_SIGNALS],
+)
+
+for signum in SIGNALS:
+    signal.signal(signum, handler)
+
+while True:
+    time.sleep(1)
+
+# vim:expandtab:smartindent:tabstop=4:softtabstop=4:shiftwidth=4:
```
Comment (clonedagain) [inline stock_scanner/hardware/datalogic/PowerScan/PowerscanService.py]: OCA doesn't use those I think

File: stock_scanner/hardware/datalogic/PowerScan/hardware.py [hunk 22]
```diff
@@ -0,0 +1,531 @@
...
+        hwd.start()
+        hwd.join()
+    except serial.SerialException, e:
+        sys.stderr.write('Error on the serial port')
+        sys.exit(0)
+
+    sys.stderr.write('User ask to shutdown the program')
+    sys.exit(0)
+
+# vim:expandtab:smartindent:tabstop=4:softtabstop=4:shiftwidth=4:
```
Comment (clonedagain) [inline stock_scanner/hardware/datalogic/PowerScan/hardware.py]: OCA doesn't use those I think

File: stock_scanner/hardware/datalogic/PowerScan/powerscan.py [hunk 23]
```diff
@@ -0,0 +1,235 @@
...
+        """ Turn on the red LED """
+        self._command("[8q")
+
+    def redledoff(self):
+        """ Turn off the red LED """
+        self._command("[9q")
+
+
+
+# vim:expandtab:smartindent:tabstop=4:softtabstop=4:shiftwidth=4:
```
Comment (clonedagain) [inline stock_scanner/hardware/datalogic/PowerScan/powerscan.py]: OCA doesn't use those I think

File: stock_scanner/load_scenario.py [hunk 24]
```diff
@@ -0,0 +1,235 @@
...
+        convert_file(cr, module, filename, idref,
+                     mode=mode, noupdate=noupdate,
+                     kind=kind, report=report, pathname=pathname)
+
+# Monkey patch Odoo's module file loading
+# To be able to load scenarios from manifest file
+openerp.tools.convert_file = scenario_convert_file
+openerp.tools.convert.convert_file = scenario_convert_file
+
+# vim:expandtab:smartindent:tabstop=4:softtabstop=4:shiftwidth=4:
```
Comment (clonedagain) [inline stock_scanner/load_scenario.py]: OCA doesn't use those I think

File: stock_scanner/models/scanner_hardware_step_history.py [hunk 25]
```diff
@@ -0,0 +1,26 @@
...
+        comodel_name='scanner.scenario.step', string='Step',
+        help='Step executed during this history line.')
+    transition_id = fields.Many2one(
+        comodel_name='scanner.scenario.transition', string='Transition',
+        help='Transition executed during this history line.')
+    message = fields.Char(
+        string='Message',
+        help='Message sent during execution of the step.')
+
+# vim:expandtab:smartindent:tabstop=4:softtabstop=4:shiftwidth=4:
```
Comment (clonedagain) [inline stock_scanner/models/scanner_hardware_step_history.py]: OCA doesn't use those I think

**Comment by clonedagain** [comment]
What is the `datalogic` directory for please?
Is it some kind of hardware driver for a specific brand or model?

**Comment by Garamotte** [comment]
The `datalogic` directory contains a client for a specific brand/model.
It was developped for a customer using the very first version of this module (on OpenERP v5.0), and wasn't updated since 2011, but I kept the files here, because it can be helpful for someone, as an example.

I removed the vim comments (travis lint check pointed this out too).

File: stock_scanner/hardware/sentinel.py (hunk 18 above)
Comment (Garamotte) [inline stock_scanner/hardware/sentinel.py]: Maybe add a second possible name, to avoid breaking already in production instances ?

**Comment by clonedagain** [comment]
@sylvain-garancher then I think you should drop the `datalogic` directory altogether.
If it's still compatible with the current version, you should spin it off to another module so that interested users can find it.
Otherwise just remove it please, and add a note in the README with the git hash.

**Comment by Garamotte** [comment]
Keeping the git hash in `README.rst` is  good idea, and it's enough to keep access to the files.
I dropped the `datalogic` directory and added a link to the directory at the current commit on GitHub.

French translation of the module updated (help sentences with periods + two forgotten constraints strings).

For the `.mo` files, I suppose you are talking about `sentinel.py` translation.
They are compiled from the `.po` files indeed, with this very simple script : https://github.com/syleam/stock-logistics-workflow/blob/8.0-add-stock_scanner-module/stock_scanner/hardware/i18n/translate.sh
I choosed to commit the `.mo` files, as it eases updates : one just has to `git pull` when updating the module, without having to think about compiling new translations.

**Comment by clonedagain** [comment]
@sylvain-garancher I see your point. We usually don't commit both the compiled/generated objects and their sources, because eventually they drift apart.
Maybe it would be acceptable if you provided a git pre-commit hook?
Another solution would be to generate the PO files at runtime. It could be as simple as running the script from the main program (but that would make it unix-only I'm afraid).

**Comment by Garamotte** [comment]
The link for the `datalogic` directory has been fixed to point to the OCA repository.

The images are now v8.0 screenshots, instead of v6.0 GTK client screenshots.

The `sentinel.py` client now accepts `.oerp_sentinelrc` or `.openerp_sentinelrc` configuration files.
If the two are available, `.openerp_sentinelrc` is loaded after, so its values overwrite others.

I think the pre-commit hook is the best solution if we have to remove the `.mo` files from the repository, but it can be forgotten too.
Does anyone else have an opinion about the `.mo` files ?

**Comment by gurneyalex** [comment]
I'm favorable to a solution where .mo files are not commited to the repository. 

Thanks for contributing this module. I deployed a version 2 years ago and it was already great at the time.

**Comment by clonedagain** [comment]
Maybe you could simply generate the MO if it's not found? or simply throw an error with the directions to generate them? I'm not sure how difficult or reliable that would be.

**Comment by damdam-s** [comment]
What about doing the sentinel part in a `pip` package on which odoo module `stock_scanner` depends.
So the translation of this app (mo files) could be generated by the install script of the setup.

**Comment by Garamotte** [comment]
For the sentinel translation files issue, I think @damdam-s idea is good, as the sentinel is not related to a specific Odoo version. If we separate the sentinel.py client in its own pip-installable python module, I think the sentinel should be moved in another git repo.
I don't know how to build mo files during install using pip though.

As this PR has not been merged yet, I took the liberty to rebase it on our module's own repository (where I added the changes made here + some other fixes).

I also removed the useful scenarii that were available in the "demo" directory, to split them in separate modules. Even if scenarii are made to be custom, it could be great to have a set of modules, each providing some "standard" scenarii.
Maybe this could become a new repository, grouping all scenarii modules ? Probably including the stock_scanner module itself ?

**Comment by gurneyalex** [comment]
@sylvain-garancher that's a great idea. 

Regarding the generation of the `.mo` files, I think this can be done in `setup.py`. See for example https://wiki.maemo.org/Internationalize_a_Python_application#Include_translations_in_your_installation

**Comment by gurneyalex** [comment]
@sylvain-garancher I'm merging this PR. We can improve it later.

**Pull request merged**